File: 123456789101112.cpp
// 12345678910111213141516...
// find digit in k posision
// with q requests. 1 request gives 1 digit
#include <iostream>
#include <math.h>
#include <vector>
using namespace std;

// Ввод числа
int Input() {
    int input;
    while (!(cin >> input) or (cin.peek() != '\n')) {
        cin.clear();
        while (cin.get() != '\n');
        cout << "Enter the number correctly!" << endl;
    }
    cin.ignore();
    return input;
}
// Ввод списка из q чисел
vector <int> Input_list(vector <int>& Query, int q) {
    for (int i = 0; i < q; i++) {
        cout << "list[" << i << "] = ";
        Query[i] = Input();
    }
    return Query;
}
// Находим нужную цифру на k-ой позиции
int find_digit(int k) {
    // Пропускаем номера всех чисел, длиной ниже искомого и узнаём длину целевого числа
    int pre_low = 0; int low = 0; int len = 0;
    while (low < k) {
        len += 1;
        pre_low = low;
        low += len * 9 * pow(10, len - 1);
    }
    low = pre_low;
    // Номер нужного подчисла в списке чисел с длиной len
    int target_number = (k - low-1) / len + pow(10, len - 1);
    // Порядок цифры в нужном подчисле
    int target_place;
    if ((k - low) % len == 0)
        target_place = len;
    else
        target_place = (k - low) % len;
    // Получение целевой цифры числа путём отбразывания лишних цифр с конца
    while (target_number >= pow(10, target_place)) {
        target_number /= 10; // Отбрасываем последнюю цифру, пока не останется нужное их количество
    }
    int digit = target_number % 10;
    return digit;
}
// Точка входа в программу
int main() {
    // Ввод числа q
    cout << "q = ";
    int q = Input();
    // Ввод списка из q чисел
    vector <int> Query(q);
    Query = Input_list(Query, q);
    // Определяем нужные цифры по запросам
    vector <int> Return(q);
    for (int i = 0; i < q; i++) {
        Return[i] = find_digit(Query[i]);
    }
    // Вывод искомых цифр
    for (int i = 0; i < q; i++) {
        cout << Return[i] << endl;
    }
    return 0;
}

File: 123456789101112.py
import math
def input_number():
    while True:
        try:
            num = int(input())
            return num
        except ValueError:
            print("Enter the number correctly!")

def input_list(q):
    query = []
    for i in range(q):
        print(f"list[{i}] = ", end = "")
        query.append(input_number())
    return query

def find_digit(k):
    # Пропускаем номера всех чисел, длиной ниже искомого и узнаём длину целевого числа
    pre_low = 0
    low = 0
    length = 0
    while low < k:
        length += 1
        pre_low = low
        low += length * 9 * (10 ** (length - 1))
    low = pre_low
    
    # Номер нужного подчисла в списке чисел с длиной length
    target_number = (k - low - 1) // length + (10 ** (length - 1))
    
    # Порядок цифры в нужном подчисле
    if (k - low) % length == 0:
        target_place = length
    else:
        target_place = (k - low) % length
    
    # Получение целевой цифры числа путём отбрасывания лишних цифр с конца
    while target_number >= (10 ** target_place):
        target_number //= 10  # Отбрасываем последнюю цифру
    
    digit = target_number % 10
    return digit
# Основные действия
def main():
    # Ввод числа q
    print("q = ", end="")
    q = input_number()
    
    # Ввод списка из q чисел
    query = input_list(q)
    
    # Определяем нужные цифры по запросам
    result = []
    for num in query:
        result.append(find_digit(num))
    
    # Вывод искомых цифр
    for digit in result:
        print(digit)
# Запуск программы
main()
